finality-aleph/src/sync/substrate/chain_status.rs

```diff
@@ -169,15 +169,31 @@ impl SubstrateChainStatus {
 impl ChainStatus<SubstrateSyncBlock, Justification> for SubstrateChainStatus {
     type Error = Error;
 
-    fn finalized_at(&self, number: BlockNumber) -> Result<Option<Justification>, Self::Error> {
+    fn finalized_at(
+        &self,
+        number: BlockNumber,
+    ) -> Result<FinalizationStatus<Justification>, Self::Error> {
+        use FinalizationStatus::*;
         let id = match self.hash_for_number(number)? {
             Some(hash) => BlockId { hash, number },
-            None => return Ok(None),
+            None => return Ok(NotFinalized),
+        };
+        let header = match self.status_of(id)? {
+            BlockStatus::Justified(justification) => {
+                return Ok(FinalizedWithJustification(justification))
+            }
+            BlockStatus::Present(id) => id,
+            _ => return Ok(NotFinalized),
         };
-        match self.status_of(id)? {
-            BlockStatus::Justified(justification) => Ok(Some(justification)),
-            _ => Ok(None),
+
+        // if we dont have justification for block at `number` level but there are finalized blocks
+        // above this level then this block is finalized.
```

Comment:
This comment would be more useful if it spelled out the actual assumption – if a block is finalized then `hash_for_number` will return the hash associated with the block rather than any fork.

Reply:
there is I hope slightly better comment now